heapster-saw/src/Verifier/SAW/Heapster/Implication.hs

```diff
@@ -5423,7 +5572,7 @@ proveVarLLVMBlocks' x ps psubst mb_bps_in mb_ps = case mbMatch mb_bps_in of
   -- permission on the left with this exact offset, length, and shape, because
   -- it would have matched some previous case, so try to eliminate a memblock
   -- and recurse
-  [nuMP| mb_bp : mb_bps |]
+  [nuMP| mb_bp : _ |]
     | [nuMP| PExpr_NamedShape _ _ nmsh _ |] <- mbMatch $ fmap llvmBlockShape mb_bp
```

Comment:
Well, I started to look at how to implement this change, and it led to a fair bit of reorganizing `proveVarLLVMBlocks`. The main changes are that it is now organized into two separate stages, represented by two separate functions, where the second stage / function takes a `MatchedMb` for the shape as an explicit argument.

Another change is that these functions now take lists of multi-bindings instead of multi-bindings of lists. The justification is that those lists are always going to be converted to pair form and matched against anyway, so we might as well do that to begin with. Plus, there are a number of recursive calls that cons or prepend to those lists, and before they had to do this with `fmap`, which converts back to function form and then again to pair form, while now this conversion has been removed. Because one of the main concerns with this change is performance, I did in fact time running Heapster on the `rust_data` example before and after the change, and it seems to improve performance slightly, on the order of .1 - .2 seconds out of 17 seconds for the entire test. So I think that's a win?